Create atlas and coordinate_space

vinaysraghavan opened the pull request:
#### Reference Issues/PRs
See #214 

#### What does this implement/fix? Explain your changes.
Fixes the distinction between coordinate spaces and brain atlases in the Brain object. Added the ability to specify atlas parcellation, including loading custom parcellations stored as .annot files.

#### Any other comments?
Should not break any prior functionality or use with loading MNI152 vs FreeSurfer and using the custom DK .annot file for the MNI152 data

vinaysraghavan committed:
Create atlas and coordinate_space #241

vinaysraghavan committed:
Updated label loader #214

Created structure for loading different labels based on coordinate space and atlas.

Still need to construct num2region dictionaries from relevant data

vinaysraghavan committed:
General atlas loader #214

While retaining custom labels for specific scenarios

vinaysraghavan committed:
Refactor more brain functions #214

Making the existing brain functions play nice with the new refactor

vinaysraghavan committed:
Coordinate space, atlas docstrings

vinaysraghavan committed:
Annotation import

vinaysraghavan committed:
Verify coordinate space

vinaysraghavan committed:
Channel offset

vinaysraghavan committed:
Handle simplified temporal ROI

vinaysraghavan committed:
Custom atlas example

vinaysraghavan committed:
Custom atlas annotation test

vinaysraghavan committed:
Add isleft defaults

gavinmischler commented on naplib/localization/freesurfer.py:
```diff
@@ -20,157 +20,31 @@
         surf_type: str = "pial",
         subject: str = "fsaverage",
+        coordinate_space: str = 'FSAverage',
+        atlas: str = '',
```
Can you change this to default to None rather than empty string?

vinaysraghavan committed:
Fix custom atlas test

gavinmischler commented on naplib/localization/freesurfer.py:
```diff
@@ -193,41 +73,47 @@ def __init__(
         self.surf_type = surf_type
         self.subject = subject
+        self.coordinate_space = coordinate_space
+        self.atlas = atlas
```
Is there input validation we can do on atlas?

vinaysraghavan replied:
Changed to check that `atlas` is either "Destrieux" or "Desikan-Killiany" or SUBJECT_DIR/SUBJECT/label/{hemi}.{atlas}.annot exists

gavinmischler commented on naplib/localization/freesurfer.py:
```diff
@@ -894,21 +800,30 @@ def reset_overlay_except(self, labels):
             Subject to use, must be a directory within ``subject_dir``
+        coordinate_space : str, default='FSAverage'
+            Coordinate space of brain vertices. Must be 'FSAverage' or 'MNI152'
+        atlas : str, default=''
```
can you also make this one default to None?

gavinmischler reviewed (commented):
Looks good! Just a couple little comments

vinaysraghavan committed:
Validate brain inputs #214

vinaysraghavan committed:
Atlas validation

vinaysraghavan committed:
Brain validation order

vinaysraghavan commented:
Still getting a little bit of unexpected behavior with `annotate_coords` and `paint_overlay`. Want to make sure everything is working cleanly first

vinaysraghavan committed:
Separate lh/rh annotation

vinaysraghavan committed:
Per hemisphere labeling

vinaysraghavan commented:
@gavinmischler 

It seems like the Hemisphere properties self.has_overlay and self.has_overlay_cells are never used. Setting self.has_overlay_cells in paint_overlay() requires calling self.zones(), which is considerably more computationally intensive than simply setting self.overlay[self.labels==self.label2num[label]] = value. 

Is there any reason to keep self.has_overlay and self.has_overlay_cells?

vinaysraghavan committed:
Fixed test result for bilateral labels

gavinmischler commented:
> It seems like the Hemisphere properties self.has_overlay and self.has_overlay_cells are never used. Setting self.has_overlay_cells in paint_overlay() requires calling self.zones(), which is considerably more computationally intensive than simply setting self.overlay[self.labels==self.label2num[label]] = value.
> 
> Is there any reason to keep self.has_overlay and self.has_overlay_cells?

Hmm I guess maybe not. I can't remember why we have those now to be honest.

vinaysraghavan committed:
Removing has_overlay

vinaysraghavan committed:
Check label type

vinaysraghavan committed:
parcellate_overlay

Merge overlay values within each atlas parcel into a single value. Meant to be used after interpolate_electrodes_onto_brain()

gavinmischler commented on naplib/localization/freesurfer.py:
```diff
@@ -1295,6 +1248,32 @@ def interpolate_electrodes_onto_brain(self, coords, values, isleft=None, k=10, m
+
+        Parameters
+        ----------
+        merge_func : callable, optional
```
I think it's better to say "callable, default=numpy.mean", because optional means that you can pass None and it will be fine, which is not the case

gavinmischler reviewed (approved):
Just a little docstring thing but looks great!

vinaysraghavan committed:
Parcellate overlay defaults
